[examples] Add example using nextjs & @mui/styles as a starter for the migration to v5 (#33005)
mui/material-ui: examples/nextjs-with-typescript-v4-migration/README.md

# Next.js with TypeScript example with @mui/styles

## How to use

Download the example [or clone the repo](https://github.com/mui/material-ui):

<!-- #default-branch-switch -->

```sh
curl https://codeload.github.com/mui/material-ui/tar.gz/master | tar -xz --strip=2  material-ui-master/examples/nextjs-with-typescript
cd nextjs-with-typescript
```

Install it and run:

```sh
npm install
npm run dev
```

or:

<!-- #default-branch-switch -->

> [![Edit on StackBlitz](https://developer.stackblitz.com/img/open_in_stackblitz.svg)](https://stackblitz.com/github/mui/material-ui/tree/master/examples/nextjs-with-typescript)
> 
> [![Edit on CodeSandbox](https://codesandbox.io/static/img/play-codesandbox.svg)](https://codesandbox.io/s/github/mui/material-ui/tree/master/examples/nextjs-with-typescript)

Reviewer: ```suggestion
[![Edit on StackBlitz](https://developer.stackblitz.com/img/open_in_stackblitz.svg)](https://stackblitz.com/github/mui/material-ui/tree/master/examples/nextjs-with-typescript-v4-migration)

[![Edit on CodeSandbox](https://codesandbox.io/static/img/play-codesandbox.svg)](https://codesandbox.io/s/github/mui/material-ui/tree/master/examples/nextjs-with-typescript-v4-migration)
```

## The idea behind the example

The project uses [Next.js](https://github.com/vercel/next.js), which is a framework for server-rendered React apps.
It includes `@mui/material` and its peer dependencies, including `emotion`, the default style engine in MUI v5. If you prefer, you can [use styled-components instead](https://mui.com/material-ui/guides/interoperability/#styled-components).
It also includes `@mui/styles`, the legacy styling solution that uses JSS as an engine.
It provides all the necessary config for configuring both emotion & JSS for server side rendering.

> The project is intended as a basic starter while you are migration your application from v4 to v5, as it allows the JSS style overrides to take presendence over the styles coming by default for the Material UI components.

Reviewer: Should we add something like
"This project demonstrates the result after [theme](https://mui.com/material-ui/migration/v5-style-changes/) and [component](https://mui.com/material-ui/migration/v5-component-changes/) breaking changes are handled."

> # Next.js with TypeScript example with @mui/styles

Reviewer 2: ```suggestion
# Material UI v5 and Next.js example with @mui/styles (in TypeScript)
```

## How to use

Download the example [or clone the repo](https://github.com/mui/material-ui):

<!-- #default-branch-switch -->

```sh
curl https://codeload.github.com/mui/material-ui/tar.gz/master | tar -xz --strip=2  material-ui-master/examples/nextjs-with-typescript
cd nextjs-with-typescript
```

Install it and run:

```sh
npm install
npm run dev
```

or:

<!-- #default-branch-switch -->

[![Edit on StackBlitz](https://developer.stackblitz.com/img/open_in_stackblitz.svg)](https://stackblitz.com/github/mui/material-ui/tree/master/examples/nextjs-with-typescript-v4-migration)

[![Edit on CodeSandbox](https://codesandbox.io/static/img/play-codesandbox.svg)](https://codesandbox.io/s/github/mui/material-ui/tree/master/examples/nextjs-with-typescript-v4-migration)

## The idea behind the example

The project uses [Next.js](https://github.com/vercel/next.js), which is a framework for server-rendered React apps.

> It includes `@mui/material` and its peer dependencies, including `emotion`, the default style engine in MUI v5. If you prefer, you can [use styled-components instead](https://mui.com/material-ui/guides/interoperability/#styled-components).

Reviewer 2: ```suggestion
It includes `@mui/material` and its peer dependencies, including `emotion`, the default style engine in Material UI v5. If you prefer, you can [use styled-components instead](https://mui.com/material-ui/guides/interoperability/#styled-components).
```

It also includes `@mui/styles`, the legacy styling solution that uses JSS as an engine.

> It provides all the necessary config for configuring both emotion & JSS for server side rendering.

Reviewer 2: ```suggestion
It provides all the necessary config for working with both Emotion and JSS for server-side rendering.
```

The project is intended as a basic starter while you are migration your application from v4 to v5, as it allows the JSS style overrides to take presendence over the styles coming by default for the Material UI components.

> It demonstrates the result after [theme](https://mui.com/material-ui/migration/v5-style-changes/) and [component](https://mui.com/material-ui/migration/v5-component-changes/) breaking changes are handled.

Reviewer 2: ```suggestion
It demonstrates what results after handling v5's breaking changes to the [theme](https://mui.com/material-ui/migration/v5-style-changes/) and [components](https://mui.com/material-ui/migration/v5-component-changes/).
```

> ## The link component

Reviewer 2: ```suggestion
## The Link component
```

Next.js has [a custom Link component](https://nextjs.org/docs/api-reference/next/link).

> The example folder provides adapters for usage with MUI.

Reviewer 2: I guess it's technically true for other MUI packages? But we don't want to encourage our users to think of MUI as a synonym for Material UI.

```suggestion
The example folder provides adapters for usage with Material UI.
```

> More information [in the documentation](https://mui.com/material-ui/guides/routing/#next-js).

Reviewer 2: To make this a complete sentence.

```suggestion
You can find more information [in the documentation](https://mui.com/material-ui/guides/routing/#next-js).
```

## What's next?

<!-- #default-branch-switch -->

You now have a working example project.

> You can head back to the documentation, continuing browsing it from the [templates](https://mui.com/material-ui/getting-started/templates/) section.

Reviewer 2: I think I would cut this. We could replace it with a section to tell people what they should do if they still need help.

Author: Done, it makes sense for the other examples, not this one.
